---
title: 'GWAS tutorial: Snp testing'
author: Patrick Breheny and Logan Harris
date: '`r format(Sys.Date(), "%B %d, %Y")`'
---

Note: Currently undergoing updates

```{r knitr_setup, include=FALSE, purl=FALSE}
library(knitr)
library(kableExtra)
set.seed(1)
knitr::opts_knit$set(aliases=c(h = 'fig.height', w = 'fig.width'))
knitr::opts_chunk$set(comment="#", message=FALSE, collapse=TRUE, cache=FALSE, tidy=FALSE, fig.align="center")
knitr::knit_hooks$set(small.mar = function(before, options, envir) {
  if (before) par(mar = c(4, 4, .1, .1))
})
```

```{r setup, include=FALSE}
library(data.table)
library(magrittr)
library(qqman)
library(snpStats)
library(dplyr)
```

# Forenote

# Adjusted

## Read in data
```{r read}
clinical <- fread("data/GWAStutorial_clinical.csv")
obj <- readRDS('data/gwas-imp.rds')
obj_pre <- readRDS('data/gwas-qc.rds')
(bim <- fread('data/GWAStutorial.bim'))
obj$genotypes
dim(obj$map)
```

## GWAS

Within the context of a GWAS study, a key assumption is that the patients are independent from one another. However, this assumption is not always met. There could be subgroups within the population which contribute to differences in allele frequencies between individuals. The presence of differences in allele frequencies due to systematic dissimilarities in ancestry is called **stratification**. Stratification violates the assumption of independence, and so an important consideration for a GWAS study is controlling for stratification alongside other population level factors. 

Since this page is being built concurrently with the analysis on population structure, we will set that aside for the moment. Later, we will return to the concept of examining the data for evidence of stratification. 

We begin our analysis with a logistic regression model which uses the predictors sex and age to study the binary outcome CAD. With the function `snp.rhs.tests()`, we will perform a large set of logistic regression tests and save the p-values in a vector called "assoc_p_vals"

```{r association_test, message = FALSE, warning = FALSE}
assoc_test <- snpStats::snp.rhs.tests(
  clinical$CAD ~ clinical$sex + clinical$age, 
  family   = "binomial",
  data     = obj$fam,
  snp.data = obj$genotypes
  # rules = rules
  )

assoc_p_vals <- p.value(assoc_test)

## Pre imputation
assoc_test_pre <- snp.rhs.tests(
  clinical$CAD ~ clinical$sex + clinical$age, 
  family   = "binomial",
  data     = obj_pre$fam,
  snp.data = obj_pre$genotypes
  )

assoc_p_vals_pre <- p.value(assoc_test_pre)

## Table to be used with FUMA/MAGMA
snp_fuma <- data.frame(
  SNP = assoc_test@snp.names,
  P = p.value(assoc_test),
  N = assoc_test@N
  ) %>%
  dplyr::left_join(bim[,c(1:2, 4)], by = c("SNP" = "V2")) %>%
  dplyr::rename(CHR = V1, POS = V4) %>%
  dplyr::select(SNP, CHR, POS, P, N)

write.table(
  snp_fuma,
  file = "data/snp_fuma.txt",
  row.names = FALSE,
  col.names = TRUE,
  quote = FALSE
)
```

### qqplot

As a first look at our results, we will examine a qq-plot. This plot compares the observed p-values from our SNP tests with the p-values we would expect if there were no associations. To make our qq-plot, we will use the package `qqman`. 

```{r assumption_test}
qqman::qq(assoc_p_vals_pre)
```

If the observed p-values follow exactly the distribution we would expect under the global null hypothesis, then all points will be plotted on the red line. 

We notice in this qq-plot that some of the p-values (plotted as points) diverge from the red line to the left. This indicates that some observed values are below what is expected. From a statistical point of view, this is pretty promising. Having some p-values that are smaller than expected indicates that there are likely to be significant results in the data. 

### Manhattan Plot

#### Data Wrangling and Default Plot

Another plot which is commonly used to examine significance of genes in specific regions is a **Manhattan plot**. After formatting our data to have readable labels, I create a Manhattan plot using the `manhattan()` function (also from the `qqman` package). 


- need to compare to pre imputation

```{r manhattan_plot}
# format data to have readable labels. 
manh_data <- data.frame(
    SNP = assoc_test@snp.names, 
    P = assoc_p_vals
  ) %>%
  left_join(bim, by = c("SNP" = "V2")) %>%
  rename(
    CHR = V1, 
    BP = V4
  )

manh_data_pre <- data.frame(
    SNP = assoc_test_pre@snp.names, 
    P = assoc_p_vals_pre
  ) %>%
  left_join(bim, by = c("SNP" = "V2")) %>%
  rename(
    CHR = V1, 
    BP = V4
  )

manhattan(manh_data, ylim = c(0, 40))
manhattan(manh_data_pre, ylim = c(0, 20))

manh_data %>% filter(P < .00000005)
manh_data_pre %>% filter(P < .00000005)
```

Here, the x-axix of the plot is divided into "bins", where each bin represents a chromosome. The y axix shows the negative, log-transformed p-values. Each of the p-values from our analysis is plotted in the bin corresponding to the chromosome of the gene in that particular test, and the height of the point directly correlates with the significance of the test.

The goal of Manhattan plots are to help identify SNPs (or a region of SNPs) that are associated with a phenotype of interest. The blue and red lines are values for $-log_{10}$ of p-values at two levels of "significance." When we do have SNPs with a p-value that exceeds these lines, we are often interested in the one that is the highest in a given region. However, we can also look closer at groups of SNPs, especially if it is the case that the trait is expected to have a complex and wide ranging levels of inheritance. The presence of such a complex trait is known as **pleiotropy** (*i.e.* many SNPs determine level of affectedness).

#### Adding annotations

```{r man_with_annotations}
manhattan(manh_data, ylim = c(0, 10), annotatePval = .000005)
```


Based on the Manhattan plots, a region of interest could be around rs9632884 on Chromosome 9. If I wanted to highlight a region of interest, I could do this to highlight the specific genes in this region, or **locus**. This will also allow us to practice the final piece of functionality from `qqman`:

```{r man_with_highlighting}
bp_center <- manh_data %>% # NB: 'bp' stands for 'base pair'
  filter(SNP == "rs9632884") %>%
  pull(BP)

bp_range <- c(-1, 1) * 100000 + bp_center

snps_highlight <- manh_data %>%
  filter(BP >= bp_range[1], BP <= bp_range[2], CHR == 9) %>%
  pull(SNP)

manhattan(
  manh_data, 
  ylim = c(0, 10), 
  annotatePval = .000005, 
  highlight = snps_highlight
)
```

The above may be a bit of an oversimplification, yet it still stands to demonstrate the basic concept of Manhattan plots and how to create them. 

In the present cases represented in the Manhattan plot of our data, we see several that exceed $-log_{10}(p)$ = 5 (or $p \leq .00005$) and one in particular exceeds $-log_{10}(p) \approx$ 7.5. (Note that the lines above can be adjusted and are default to the `manhattan()` function.) 

In practice, we would ideally set significance levels prior to analysis or use some correction to avoid bias. For the sake of demonstration, we will carry the present analysis forward as we consider if statistical relevance translates into any coherent biological significance (sees part 4).

# For futher reading (optional)

Out of curiosity, I want to examine what happens if we do NOT adjust using imputation: 

## Read in data
```{r read_uncleaned}
obj_unc <- read.plink('data/GWAStutorial')
```

## GWAS

```{r association_test_unc, message = FALSE, warning = FALSE}
assoc_test_unc <- snp.rhs.tests(
  clinical$CAD ~ clinical$sex + clinical$age, 
  family   = "binomial",
  data     = obj_unc$fam,
  snp.data = obj_unc$genotypes
)

assoc_p_vals_unc <- p.value(assoc_test_unc)

## List of SNPs for analysis later
write.table(
  manh_data %>% filter(P <= .0005) %>% pull(SNP),
  file = "data/snp_list.txt",
  row.names = FALSE,
  col.names = FALSE,
  quote = FALSE
)
```

It drops some data and throws a warning, but let's see how the plots compare.

### qqplot

```{r assumption_test_unc}
qq(assoc_p_vals_unc)
```

Looks pretty similar to with the cleaned data, probably because this dataset was already relatively clean.

### Manhattan Plot

```{r mahnattan_plot}
manh_data_unc <- data.frame(
    SNP = assoc_test_unc@snp.names, 
    P = assoc_p_vals_unc
  ) %>%
  left_join(bim, by = c("SNP" = "V2")) %>%
  rename(
    CHR = V1, 
    BP = V4
  )

manhattan(manh_data_unc, ylim = c(0,10), annotatePval = .000005)
```

This is interesting. We see the results vary some, but the main bits of significance are nearly the same as with the cleaned data. This is likely due to the fact that by and large the data was pretty clean going into the process. In practice we should not expect this to be the case and should always begin with steps 1 and 2 of pre-processing the data.

# Attempt using plink (1.9)

PLINK is a great tool, but it didn't necessarily like the naming of the original plink object. Thus, below I recreate a plink that it agrees well with. 

```{r plink_creation, message = FALSE}
write.plink(
  file.base = "data/qc_data",
  snps = obj$genotypes,
  pedigree = obj$fam$pedigree,
  id = obj$fam$member,
  father = obj$fam$father,
  mother = obj$fam$mother,
  sex = clinical$sex,
  phenotype = clinical$CAD + 1,
  chromosome = obj$map$chromosome,
  genetic.distance = obj$map$cM,
  position = obj$map$position,
  allele.1 = obj$map$allele.1,
  allele.2 = obj$map$allele.2
)
```


The next step is also to create a covariate table.

```{r covariate_creation}
covar <- data.frame(
  FID = obj$fam$pedigree,
  IID = obj$fam$pedigree,
  sex = clinical$sex,
  age = clinical$age
)
head(covar)
write.table(
  covar,
  file = "data/covar.txt",
  row.names = FALSE,
  quote = FALSE
)
```

Then comes running the commands for plink. The first thing we have to do is have it take the plink data and make a new binary fileset. The reasons for which I am not entirely familiar, but it solved the error of ".bim file has a split chromosome" so I'll take it! Finally, we are able to then run and read the results into R. 

```{r plink, message = FALSE}
cd <- getwd()
system(paste("cd", cd))
system("plink --make-bed -bfile data/qc_data --out data/qc_data_remedy")
system("plink --bfile data/qc_data_remedy --logistic --adjust --out data/assoc_results")
adj_res <- fread("data/assoc_results.assoc.logistic")
adj_res <- na.omit(adj_res)
```

```{r plink_plots}
qq(adj_res$P)

manh_data_plink <- 
  data.frame(
    SNP = adj_res$SNP, 
    P = adj_res$P
  ) %>%
  left_join(bim, by = c("SNP" = "V2")) %>%
  rename(
    CHR = V1, 
    BP = V4
  )

manhattan(manh_data_plink, ylim = c(0,10), annotatePval = .000005)
```

Here we have a slight difference, likely due to the exclusion of covariates for the pure fact that I have been unable to overcome an error and have spent far too much time trying to trouble shoot already. The battle is to be continued.