---
title: 'GWAS tutorial: An Introduction'
author: Patrick Breheny, Anna Reisetter, Logan Harris, and Tabitha Peter
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    toc: false
---

```{r knitr_setup, include=FALSE, purl=FALSE}
library(knitr)
library(kableExtra)
set.seed(1)
knitr::opts_knit$set(aliases=c(h = 'fig.height', w = 'fig.width'))
knitr::opts_chunk$set(comment="#", message=FALSE, collapse=TRUE, cache=FALSE, tidy=FALSE, fig.align="center")
knitr::knit_hooks$set(small.mar = function(before, options, envir) {
  if (before) par(mar = c(4, 4, .1, .1))
})
```

# Welcome!

This tutorial is a continuing work in progress, but we are excited you are using this resource to get started with Genome Wide Association Studies.

This tutorial was inspired by similar work in [Reed2015](https://dx.doi.org/10.1002/sim.6605) and [Marees2017](https://onlinelibrary.wiley.com/doi/abs/10.1002/mpr.1608). Unfortunately, the software used in the 2015 tutorial is a bit out of date now -- several R packages have changed or are no longer available -- and it needed an update.

This tutorial is broken down into four sections: 1) Data formats, summary statistics and quality control, 2) Imputation and population structure, 3) SNP Testing, and 4) Post analysis and biological relevance.

## [Data Formats, Summary Statistics, and QC](quality_control.html)

This section covers the basics from different file types to initial things to look for in the data and how to exclude certain parts of the data that would muddy our end results.

## [Imputation](imputation.html)

The next section covers some important considerations that should eb taken into account before completing the analysis but which are not necessarily covered under quality control. Imputation is a method underwhich missing data is replaced using a logical method. Population structure helps us take into account possible confounders that may exist in our dataset that would incorrectly inflate results such as familial relations.

## [Population Structure](population_structure.html)

## [SNP Testing](gwas-3.html)

SNP testing gets down the results we are looking for: associated loci for our phenotypes of interest. This is the easy part if we complete the prior two sections correctly.

## [Post Analysis](gwas-4.html)

Finally, we attempt to take the results and turn them into something meaningful. For instance, I could tell you some SNP, but it will have no mean to you unless you have spent time studying it. On the other hand, if I tell you that a variant of a SNP is associated with improper functioning of chloride channels, that will have a great deal more significance.
